File: No.8_spine.m
% Use Spine interpolation to realize outline of hands
% Extra requirement:
% 1) try to use both Cartesian and polar coordinates
% 2) try to use both spline and cubic Hermite interpolation

%----------- put your hand on screens and click around -----------%
figure('position',get(0,'screensize'))
axes('position',[0 0 1 1])
[x,y] = ginput;
%------------ spline and cubic Hermite interpolation -------------%
n = length(x);
s = (1:n)';
t = (1:.05:n)';
u = splinetx(s,x,t);
v = splinetx(s,y,t);
a = pchiptx(s,x,t);
b = pchiptx(s,y,t);
%---------- spline interpolation in polar coordinates -----------%
r = sqrt(x.^2+y.^2);
ta = atan2(y,x);
n = length(ta);
s = (1:n)';
t = (1:.05:n)';
c = splinetx(s,ta,t);
d = splinetx(s,r,t);
%---------------------------- plot -----------------------------%
clf reset
subplot(2,1,1)
plot(x,y,'.','Color','g');
hold on
plot(u,v,'b');
plot(a,b,'r');
title('Outline of hand in Cartesian coordinate')
legend('input point','spline','cubic Hermite')
subplot(2,1,2)
polar(c,d); 
hold on                         
polar(c,d,'.');
title('Outline of hand in polar coordinate')

File: No.8_spine.py
import numpy as np
import matplotlib.pyplot as plt
from scipy import interpolate

# Since realization of mouse clicking for outline of hand in Python requires
# much unrelated effort, here we simply read the coordinate from files
x, y = np.loadtxt(handOutline, delimiter = ',', unpack = True)

#------------ spline and cubic Hermite interpolation -------------#
n = np.size(x)
s = np.transpose(np.arange(1, n, 1))
t = np.transpose(np.arange(1, n, 0.05))

u = np.interpolate.splrep(s,x)
u = np.interpolate.splev(s,u)
v = np.interpolate.splrep(s,y)
v = np.interpolate.splev(s,v)

a = np.interpolate.interp1d(s,x, kind = 'cubic')
b = np.interpolate.interp1d(s,y, kind = 'cubic')
#---------- spline interpolation in polar coordinates -----------#
r = np.sqrt(np.power(x,2)+np.power(y,2))
ta = np.arctan2(y,x)
n = np.size(ta)
s = np.transpose(np.arange(1, n, 1))
t = np.transpose(np.arange(1, n , 0.05))

c = np.interpolate.splrep(s,ta)
c = np.interpolate.splev(s,c)
d = np.interpolate.splrep(s,r)
d = np.interpolate.splev(s,d)
#---------------------------- plot -----------------------------#

plt.subplot(2,1,1)
plt.plot(x,y,'.g')
plt.plot(u,v,'b')
plt.plot(a,b,'r')
plt.title('Outline of hand in Cartesian coordinate')
plt.legend('input point','spline','cubic Hermite')

plt.subplot(2,1,2)
plt.polar(c,d)                 
plt.polar(c,d,'.')
plt.title('Outline of hand in polar coordinate')

plt.show()
